### DISCUSSION TITLE
Is it possible to remove the caption from HTML tables in PDF output?

### DISCUSSION THREAD

**Discussion by lballabio** [quarto-dev/quarto-cli#10677] [Q&A] [answered]
### Description

Quarto 1.5.56, Mac OS 14.6.1.

This sample qmd document shows the issue in a minimal way, but I came across it while using a Jupyter notebook so it'd be great if the solution (if any) also worked there.

````qmd
---
title: "table demo"
jupyter: python3
---

```{python}
import pandas as pd

df = pd.DataFrame(
  [
    [1.1834, "a"],
    [2.4552, "b"],
    [3.8457, "c"],
  ],
  columns = ['x', 'alpha']
)
```

```{python}
df
```

```{python}
df.style.format({"x": "{:.2f}"})
```
````

The second and third cells cause a table to be generated in the PDF output.  However, the styled one from the third cell also has a caption added (I'm guessing that's because it's an HTML table?)

<img width="817" alt="Screenshot of the generated PDF" src="https://github.com/user-attachments/assets/f609c5ac-a9ee-4887-b2f6-df38c082c0cd">

Is there any way to prevent that "Table 2" from being added, or to add a "Table 1"?  Thanks!

**Comment by cscheid** [comment, marked as answer]
> (I'm guessing that's because it's an HTML table?)

Yes, but not exactly. Here's what's going on, and how to learn to diagnose these issues in the future.

First, do this to your document front matter:

```
---
title: "table demo"
jupyter: python3
keep-md: true
keep-tex: true
format: pdf
---
```

My local file is called `10677.qmd`, so:

- `keep-md: true` will keep a file called `10677.pdf.md` in your file system.
- `keep-tex: true` will keep `10677.tex`

## `10677.pdf.md`

`df` always emits HTML tables, apparently. The relevant outputs are:


### `df`

````
::: {.cell execution_count=2}
``` {.python .cell-code}
df
```

::: {.cell-output .cell-output-display execution_count=8}

```{=html}
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x</th>
      <th>alpha</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1.1834</td>
      <td>a</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2.4552</td>
      <td>b</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3.8457</td>
      <td>c</td>
    </tr>
  </tbody>
</table>
</div>
```
:::
:::
````

### `df.style.format({"x": "{:.2f}"})`

````
::: {.cell execution_count=3}
``` {.python .cell-code}
df.style.format({"x": "{:.2f}"})
```

::: {.cell-output .cell-output-display execution_count=9}

```{=html}
<style type="text/css">
</style>
<table id="T_a0997">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_a0997_level0_col0" class="col_heading level0 col0" >x</th>
      <th id="T_a0997_level0_col1" class="col_heading level0 col1" >alpha</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_a0997_level0_row0" class="row_heading level0 row0" >0</th>
      <td id="T_a0997_row0_col0" class="data row0 col0" >1.18</td>
      <td id="T_a0997_row0_col1" class="data row0 col1" >a</td>
    </tr>
    <tr>
      <th id="T_a0997_level0_row1" class="row_heading level0 row1" >1</th>
      <td id="T_a0997_row1_col0" class="data row1 col0" >2.46</td>
      <td id="T_a0997_row1_col1" class="data row1 col1" >b</td>
    </tr>
    <tr>
      <th id="T_a0997_level0_row2" class="row_heading level0 row2" >2</th>
      <td id="T_a0997_row2_col0" class="data row2 col0" >3.85</td>
      <td id="T_a0997_row2_col1" class="data row2 col1" >c</td>
    </tr>
  </tbody>
</table>
```

:::
:::
````

Quarto's [table processing](https://quarto.org/docs/authoring/tables.html#html-tables) kicks in for both tables, and produces this LaTeX output for each:

### `df`

````
\begin{longtable}[]{@{}lll@{}}
\toprule\noalign{}
& x & alpha \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
0 & 1.1834 & a \\
1 & 2.4552 & b \\
2 & 3.8457 & c \\
\end{longtable}
````

### `df.style.format({"x": "{:.2f}"})`

````
\begin{longtable}[]{@{}lll@{}}
\caption{}\label{T_a0997}\tabularnewline
\toprule\noalign{}
~ & x & alpha \\
\midrule\noalign{}
\endfirsthead
\toprule\noalign{}
~ & x & alpha \\
\midrule\noalign{}
\endhead
\bottomrule\noalign{}
\endlastfoot
0 & 1.18 & a \\
1 & 2.46 & b \\
2 & 3.85 & c \\
\end{longtable}
````

The important bit to note is the presence of a label in the second table. Quarto's processing identifies the table ID attribute in the HTML, and converts it to a label in the LaTeX output. This is "working as intended".

Pandas's styled writer needs an ID in order for its CSS to work, but that ID then causes LaTeX to think you want to reference the table. The solution here is for you to write a filter that detects that table and removes the ID. Like this:

````
---
title: "table demo"
jupyter: python3
keep-md: true
keep-tex: true
format: pdf
filters:
  - remove_table_id.lua
---

...
````

````lua
-- remove_table_id.lua
function Table(tbl)
  -- this will remove all identifiers.
  -- You should probably ensure that only the desired identifiers are removed when you
  -- write your own filter.
  tbl.identifier = ""
  return tbl
end
````

Result:

<img width="693" alt="image" src="https://github.com/user-attachments/assets/678ae3bd-65b1-4b50-9e9c-51a0abdfeeeb">

**Reply by lballabio** [reply]
Thanks for your time and effort!  That was quite the guided tour you gave me.  Much appreciated.
